kv/kvclient/kvcoord: TestBadRequest failed [use of Span after Finish]

kv/kvclient/kvcoord.TestBadRequest [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/5408869?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/5408869?buildTab=artifacts#/) on master @ [18035843c5e82dac95834dee7edc6c9dfdfc2afc](https://github.com/cockroachdb/cockroach/commits/18035843c5e82dac95834dee7edc6c9dfdfc2afc):


```
=== RUN   TestBadRequest
    test_log_scope.go:79: test logs captured to: /artifacts/tmp/_tmp/bf9904731185fc1f0815da569e14c3f5/logTestBadRequest854780260
    test_log_scope.go:80: use -show-logs to present logs inline
```
<p>Parameters: <code>TAGS=bazel,gss,deadlock</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/kv
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestBadRequest.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

```
=== RUN   TestBadRequest
    test_log_scope.go:79: test logs captured to: /artifacts/tmp/_tmp/bf9904731185fc1f0815da569e14c3f5/logTestBadRequest854780260
    test_log_scope.go:80: use -show-logs to present logs inline
panic: use of Span after Finish. Span: dist sender. Finish previously called at: <stack not captured. Set debugUseAfterFinish> [recovered]
	panic: use of Span after Finish. Span: dist sender. Finish previously called at: <stack not captured. Set debugUseAfterFinish>

goroutine 7674233 [running]:
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).Recover(0xc00d314510, {0x5848a18, 0xc01400f680})
	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:251 +0x94
panic({0x3c783a0, 0xc00d9b8470})
	GOROOT/src/runtime/panic.go:1038 +0x215
github.com/cockroachdb/cockroach/pkg/util/tracing.(*Span).detectUseAfterFinish(0x5848a18)
	github.com/cockroachdb/cockroach/pkg/util/tracing/span.go:186 +0x134
github.com/cockroachdb/cockroach/pkg/util/tracing.(*Span).Tracer(...)
	github.com/cockroachdb/cockroach/pkg/util/tracing/span.go:229
github.com/cockroachdb/cockroach/pkg/util/tracing.EnsureForkSpan({0x5848a18, 0xc01400f7d0}, 0xc00e902690, {0xc005924120, 0x29})
	github.com/cockroachdb/cockroach/pkg/util/tracing/tracer.go:1554 +0x9a
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).handleClosedTimestampUpdateRaftMuLocked(0xc001759c00, {0x5848a18, 0xc01400f7d0}, {0xc00c2f7308, 0x8ccc8a0, 0xc0})
	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/replica_rangefeed.go:698 +0x314
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).registerWithRangefeedRaftMuLocked(0xc001759c00, {0x5848a18, 0xc01400f7d0}, {{0xc009a9c8e0, 0x1, 0x8}, {0xc009a9c8e8, 0x1, 0x8}}, {0x16f692109d434629, ...}, ...)
	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/replica_rangefeed.go:428 +0x755
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).rangeFeedWithRangeID(0xc001759c00, 0x58acab0, 0xc01597c900, {0x7f511ec55d98, 0xc0040e0740})
	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/replica_rangefeed.go:228 +0x5d9
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).RangeFeed(...)
	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/replica_rangefeed.go:139
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Store).RangeFeed(0xc001140000, 0xc01597c900, {0x58acab0, 0xc0040e0740})
	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/store.go:3088 +0xf7
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Stores).RangeFeed(0xc0024c7868, 0xc01597c900, {0x58acab0, 0xc0040e0740})
	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/stores.go:216 +0xe8
github.com/cockroachdb/cockroach/pkg/server.(*Node).RangeFeed(0x58acab0, 0xc0004ea800, {0x58acab0, 0xc0040e0740})
	github.com/cockroachdb/cockroach/pkg/server/node.go:1225 +0x38
github.com/cockroachdb/cockroach/pkg/rpc.internalClientAdapter.RangeFeed.func2.1({0xc0040e0700, 0x5848a18}, {0xc01400f680, 0x43e3e55})
	github.com/cockroachdb/cockroach/pkg/rpc/pkg/rpc/context.go:934 +0x7b
github.com/cockroachdb/cockroach/pkg/util/tracing/grpcinterceptor.StreamServerInterceptor.func1({0x42f7dc0, 0xc00c1ca000}, {0x587e7d0, 0xc0040e0700}, 0x7d3b230, 0xc005b1c980)
	github.com/cockroachdb/cockroach/pkg/util/tracing/grpcinterceptor/grpc_interceptor.go:163 +0x6c2
github.com/cockroachdb/cockroach/pkg/rpc.serverStreamInterceptorsChain.run.func1({0x42f7dc0, 0xc00c1ca000}, {0x587e7d0, 0xc0040e0700})
	github.com/cockroachdb/cockroach/pkg/rpc/pkg/rpc/context.go:825 +0x5b
github.com/cockroachdb/cockroach/pkg/rpc.NewServerEx.func4({0x42f7dc0, 0xc00c1ca000}, {0x587e7d0, 0xc0040e0700}, 0xc008ccc7e0, 0xc00b91ed00)
	github.com/cockroachdb/cockroach/pkg/rpc/pkg/rpc/context.go:273 +0x74
github.com/cockroachdb/cockroach/pkg/rpc.serverStreamInterceptorsChain.run.func1({0x42f7dc0, 0xc00c1ca000}, {0x587e7d0, 0xc0040e0700})
	github.com/cockroachdb/cockroach/pkg/rpc/pkg/rpc/context.go:828 +0x83
github.com/cockroachdb/cockroach/pkg/rpc.kvAuth.streamInterceptor({{{0xc00d314530}}}, {0x42f7dc0, 0xc00c1ca000}, {0x587e7d0, 0xc0040e0700}, 0x7d3b230, 0xc00b91ed00)
	github.com/cockroachdb/cockroach/pkg/rpc/pkg/rpc/auth.go:98 +0x1c3
github.com/cockroachdb/cockroach/pkg/rpc.serverStreamInterceptorsChain.run.func1({0x42f7dc0, 0xc00c1ca000}, {0x587e7d0, 0xc0040e0700})
	github.com/cockroachdb/cockroach/pkg/rpc/pkg/rpc/context.go:828 +0x83
github.com/cockroachdb/cockroach/pkg/rpc.NewServerEx.func2.1({0xc00d314510, 0x0})
	github.com/cockroachdb/cockroach/pkg/rpc/pkg/rpc/context.go:242 +0x2d
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunTaskWithErr(0xc00d314510, {0x5848a18, 0xc01400f680}, {0x4286d4, 0x943446}, 0xc001ba2e58)
	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:344 +0xdd
github.com/cockroachdb/cockroach/pkg/rpc.NewServerEx.func2({0x42f7dc0, 0xc00c1ca000}, {0x587e7d0, 0xc0040e0700}, 0x7d3b230, 0xc00b91ed00)
	github.com/cockroachdb/cockroach/pkg/rpc/pkg/rpc/context.go:241 +0xf3
github.com/cockroachdb/cockroach/pkg/rpc.serverStreamInterceptorsChain.run.func1({0x42f7dc0, 0xc00c1ca000}, {0x587e7d0, 0xc0040e0700})
	github.com/cockroachdb/cockroach/pkg/rpc/pkg/rpc/context.go:828 +0x83
github.com/cockroachdb/cockroach/pkg/rpc.serverStreamInterceptorsChain.run({0xc00deaad00, 0x4, 0x4}, {0x42f7dc0, 0xc00c1ca000}, {0x587e7d0, 0xc0040e0700}, 0x7d3b230, 0xc005b1c980)
	github.com/cockroachdb/cockroach/pkg/rpc/pkg/rpc/context.go:830 +0x142
github.com/cockroachdb/cockroach/pkg/rpc.internalClientAdapter.RangeFeed.func2()
	github.com/cockroachdb/cockroach/pkg/rpc/pkg/rpc/context.go:938 +0x1f5
created by github.com/cockroachdb/cockroach/pkg/rpc.internalClientAdapter.RangeFeed
	github.com/cockroachdb/cockroach/pkg/rpc/pkg/rpc/context.go:931 +0x35c
I220608 06:41:54.108340 1 (gostd) testmain.go:498  [-] 1  Test //pkg/kv/kvclient/kvcoord:kvcoord_test exited with error code 2
```

--- a/pkg/rpc/context.go
+++ b/pkg/rpc/context.go
@@ -902,57 +902,87 @@ var rangefeedStreamInfo = &grpc.StreamServerInfo{
 func (a internalClientAdapter) RangeFeed(
 	ctx context.Context, args *roachpb.RangeFeedRequest, opts ...grpc.CallOption,
 ) (roachpb.Internal_RangeFeedClient, error) {
-	// Run the client interceptors.
-
 	// Create a pipe between the server-side sender and the client-side receiver.
-	// This pipe will be returned at the bottom of the interceptor chain. The
-	// interceptors might wrap it in their own ClientStream implementations, so it
-	// might not be the stream that we ultimately return to callers. But we'll
-	// connect the server-side handler to it through a rangeFeedServerAdapter.
-	streamPipe := makeRangeFeedPipe(grpcutil.NewLocalRequestContext(ctx))
-	streamer := func(
-		ctx context.Context, desc *grpc.StreamDesc, cc *grpc.ClientConn, method string, opts ...grpc.CallOption,
-	) (grpc.ClientStream, error) {
-		return streamPipe, nil
-	}
-	// clientStream might be rangeFeedPipe, or it might be another
-	// grpc.ClientStream implementation that wraps it. We're going to return
-	// clientStream to the caller (wrapped in a rangeFeedClient),
-	clientStream, err := a.clientStreamInterceptors.run(ctx, rangeFeedDesc, nil /* ClientConn */, rangefeedMethodName, streamer, opts...)
-	if err != nil {
-		return nil, err
-	}
+	// On the server side, this pipe will be possibly wrapped by server-side
+	// interceptors providing their own implementation of grpc.ServerStream, and
+	// then it will be in turn wrapped by a rangeFeedServerAdapter before being
+	// passed to the RangeFeed RPC handler (i.e. Node.RangeFeed).
+	//
+	// On the client side, this pipe will be returned at the bottom of the
+	// interceptor chain. The client-side interceptors might wrap it in their own
+	// ClientStream implementations, so it might not be the stream that we
+	// ultimately return to callers. Similarly, the server-side interceptors might
+	// wrap it before passing it to the RPC handler.
+	//
+	// The flow of data through the pipe, from producer to consumer:
+	//   RPC handler (i.e. Node.RangeFeed) ->
+	//    -> rangeFeedServerAdapter
+	//    -> grpc.ServerStream implementations provided by server-side interceptors
+	//    -> rfPipe
+	//    -> grpc.ClientStream implementations provided by client-side interceptors
+	//    -> rangeFeedClientAdapter
+	//    -> RPC caller
+	rfPipe := newRangeFeedPipe(grpcutil.NewLocalRequestContext(ctx))
 
-	// Mark this as originating locally.
+	// Mark this request as originating locally.
 	args.AdmissionHeader.SourceLocation = roachpb.AdmissionHeader_LOCAL
 
 	// Spawn a goroutine running the server-side handler. This goroutine
 	// communicates with the client stream through rangeFeedPipe.
 	go func() {
-		rfAdapter := rangeFeedServerAdapter{rangeFeedPipe: streamPipe}
+		// Handler adapts the ServerStream to the typed interface expected by the
+		// RPC handler (Node.RangeFeed). `stream` might be `rfPipe` which we
+		// pass to the interceptor chain below, or it might be another
+		// implementation of `ServerStream` that wraps it; in practice it will be
+		// tracing.grpcinterceptor.StreamServerInterceptor.
 		handler := func(srv interface{}, stream grpc.ServerStream) error {
-			return a.server.RangeFeed(args, rfAdapter)
+			return a.server.RangeFeed(args, rangeFeedServerAdapter{ServerStream: stream})
 		}
-		// Run the server interceptors, which will bottom out by running the actual
-		// handlers. This call is blocking.
-		err = a.serverStreamInterceptors.run(a.server, rfAdapter, rangefeedStreamInfo, handler)
+		// Run the server interceptors, which will bottom out by running `handler`
+		// (defined just above), which runs Node.RangeFeed (our RPC handler).
+		// This call is blocking.
+		err := a.serverStreamInterceptors.run(a.server, rfPipe, rangefeedStreamInfo, handler)
 		if err == nil {
 			err = io.EOF
 		}
-		rfAdapter.errC <- err
+		rfPipe.errC <- err
 	}()
 
-	return rangeFeedClient{clientStream}, nil
+	// Run the client-side interceptors, which produce a gprc.ClientStream.
+	// clientStream might end up being rfPipe, or it might end up being another
+	// grpc.ClientStream implementation that wraps it.
+	//
+	// NOTE: For actual RPCs, going to a remote note, there's a tracing client
+	// interceptor producing a tracing.grpcinterceptor.tracingClientStream
+	// implementation of ClientStream. That client interceptor does not run for
+	// these local requests handled by the internalClientAdapter (as opposed to
+	// the tracing server interceptor, which does run).
+	clientStream, err := a.clientStreamInterceptors.run(ctx, rangeFeedDesc, nil /* ClientConn */, rangefeedMethodName,
+		// This function runs at the bottom of the client interceptor stack,
+		// pretending to actually make an RPC call. We don't make any calls, but
+		// return the pipe on which messages from the server will come.
+		func(
+			ctx context.Context, desc *grpc.StreamDesc, cc *grpc.ClientConn, method string, opts ...grpc.CallOption,
+		) (grpc.ClientStream, error) {
+			return rfPipe, nil
+		},
+		opts...)
+	if err != nil {
+		return nil, err
+	}
+
+	return rangeFeedClientAdapter{clientStream}, nil
 }
 
-// rangeFeedClient wraps a ClientStream and implements roachpb.Internal_RangeFeedClient.
-type rangeFeedClient struct {
+// rangeFeedClientAdapter adapts an untyped ClientStream to the typed
+// roachpb.Internal_RangeFeedClient used by the rangefeed RPC client.
+type rangeFeedClientAdapter struct {
 	grpc.ClientStream
 }
 
-var _ roachpb.Internal_RangeFeedClient = rangeFeedClient{}
+var _ roachpb.Internal_RangeFeedClient = rangeFeedClientAdapter{}
 
-func (x rangeFeedClient) Recv() (*roachpb.RangeFeedEvent, error) {
+func (x rangeFeedClientAdapter) Recv() (*roachpb.RangeFeedEvent, error) {
 	m := new(roachpb.RangeFeedEvent)
 	if err := x.ClientStream.RecvMsg(m); err != nil {
 		return nil, err
@@ -960,17 +990,23 @@ func (x rangeFeedClient) Recv() (*roachpb.RangeFeedEvent, error) {
 	return m, nil
 }
 
-// rangeFeedPipe is a pipe of *RangeFeedEvent that implements the grpc.ClientStream interface
+// rangeFeedPipe is a (uni-directional) pipe of *RangeFeedEvent that implements
+// the grpc.ClientStream and grpc.ServerStream interfaces.
 type rangeFeedPipe struct {
 	ctx   context.Context
 	respC chan interface{}
 	errC  chan error
 }
 
-var _ grpc.ClientStream = rangeFeedPipe{}
+var _ grpc.ClientStream = &rangeFeedPipe{}
+var _ grpc.ServerStream = &rangeFeedPipe{}
 
-func makeRangeFeedPipe(ctx context.Context) rangeFeedPipe {
-	return rangeFeedPipe{
+// newRangeFeedPipe creates a rangeFeedPipe. The pipe is returned as a pointer
+// for convenience, because it's used as a grpc.ClientStream and
+// grpc.ServerStream, and these interfaces are implemented on the pointer
+// receiver.
+func newRangeFeedPipe(ctx context.Context) *rangeFeedPipe {
+	return &rangeFeedPipe{
 		ctx:   ctx,
 		respC: make(chan interface{}, 128),
 		errC:  make(chan error, 1),
@@ -978,20 +1014,35 @@ func makeRangeFeedPipe(ctx context.Context) rangeFeedPipe {
 }
 
 // grpc.ClientStream methods.
-func (rangeFeedPipe) Header() (metadata.MD, error) { panic("unimplemented") }
-func (rangeFeedPipe) Trailer() metadata.MD         { panic("unimplemented") }
-func (rangeFeedPipe) CloseSend() error             { panic("unimplemented") }
+func (*rangeFeedPipe) Header() (metadata.MD, error) { panic("unimplemented") }
+func (*rangeFeedPipe) Trailer() metadata.MD         { panic("unimplemented") }
+func (*rangeFeedPipe) CloseSend() error             { panic("unimplemented") }
 
 // grpc.ServerStream methods.
-func (rangeFeedPipe) SetHeader(metadata.MD) error  { panic("unimplemented") }
-func (rangeFeedPipe) SendHeader(metadata.MD) error { panic("unimplemented") }
-func (rangeFeedPipe) SetTrailer(metadata.MD)       { panic("unimplemented") }
+func (*rangeFeedPipe) SetHeader(metadata.MD) error  { panic("unimplemented") }
+func (*rangeFeedPipe) SendHeader(metadata.MD) error { panic("unimplemented") }
+func (*rangeFeedPipe) SetTrailer(metadata.MD)       { panic("unimplemented") }
 
-// grpc.Stream methods.
-func (p rangeFeedPipe) Context() context.Context  { return p.ctx }
-func (rangeFeedPipe) SendMsg(m interface{}) error { panic("unimplemented") }
+// Common grpc.{Client,Server}Stream methods.
+func (p *rangeFeedPipe) Context() context.Context { return p.ctx }
 
-func (p rangeFeedPipe) RecvMsg(m interface{}) error {
+// SendMsg is part of the grpc.ServerStream interface. It is also part of the
+// grpc.ClientStream interface but, in the case of the RangeFeed RPC (which is
+// only server-streaming, not bi-directional), only the server sends.
+func (p *rangeFeedPipe) SendMsg(m interface{}) error {
+	select {
+	case p.respC <- m:
+		return nil
+	case <-p.ctx.Done():
+		return p.ctx.Err()
+	}
+}
+
+// RecvMsg is part of the grpc.ClientStream interface. It is also technically
+// part of the grpc.ServerStream interface but, in the case of the RangeFeed RPC
+// (which is only server-streaming, not bi-directional), only the client
+// receives.
+func (p *rangeFeedPipe) RecvMsg(m interface{}) error {
 	out := m.(*roachpb.RangeFeedEvent)
 	msg, err := p.recvInternal()
 	if err != nil {
@@ -1001,7 +1052,8 @@ func (p rangeFeedPipe) RecvMsg(m interface{}) error {
 	return nil
 }
 
-func (p rangeFeedPipe) recvInternal() (interface{}, error) {
+// recvInternal is the implementation of RecvMsg.
+func (p *rangeFeedPipe) recvInternal() (interface{}, error) {
 	// Prioritize respC. Both channels are buffered and the only guarantee we
 	// have is that once an error is sent on errC no other events will be sent
 	// on respC again.
@@ -1019,33 +1071,28 @@ func (p rangeFeedPipe) recvInternal() (interface{}, error) {
 	}
 }
 
-func (p rangeFeedPipe) sendInternal(e interface{}) error {
-	select {
-	case p.respC <- e:
-		return nil
-	case <-p.ctx.Done():
-		return p.ctx.Err()
-	}
-}
-
+// rangeFeedServerAdapter adapts an untyped ServerStream to the typed
+// roachpb.Internal_RangeFeedServer interface, expected by the RangeFeed RPC
+// handler.
 type rangeFeedServerAdapter struct {
-	rangeFeedPipe
+	grpc.ServerStream
 }
 
 var _ roachpb.Internal_RangeFeedServer = rangeFeedServerAdapter{}
 
 // roachpb.Internal_RangeFeedServer methods.
 func (a rangeFeedServerAdapter) Recv() (*roachpb.RangeFeedEvent, error) {
-	e, err := a.recvInternal()
+	out := &roachpb.RangeFeedEvent{}
+	err := a.RecvMsg(out)
 	if err != nil {
 		return nil, err
 	}
-	return e.(*roachpb.RangeFeedEvent), nil
+	return out, nil
 }
 
-// roachpb.Internal_RangeFeedServer methods.
+// Send implement the roachpb.Internal_RangeFeedServer interface.
 func (a rangeFeedServerAdapter) Send(e *roachpb.RangeFeedEvent) error {
-	return a.sendInternal(e)
+	return a.ServerStream.SendMsg(e)
 }
 
 // IsLocal returns true if the given InternalClient is local.

--- a/pkg/rpc/context_test.go
+++ b/pkg/rpc/context_test.go
@@ -225,6 +225,7 @@ var _ roachpb.InternalServer = &internalServer{}
 type internalServer struct {
 	// rangeFeedEvents are returned on RangeFeed() calls.
 	rangeFeedEvents []roachpb.RangeFeedEvent
+	serverStream    roachpb.Internal_RangeFeedServer
 }
 
 func (*internalServer) Batch(
@@ -242,6 +243,7 @@ func (*internalServer) RangeLookup(
 func (s *internalServer) RangeFeed(
 	_ *roachpb.RangeFeedRequest, stream roachpb.Internal_RangeFeedServer,
 ) error {
+	s.serverStream = stream
 	for _, ev := range s.rangeFeedEvents {
 		evCpy := ev
 		if err := stream.Send(&evCpy); err != nil {
@@ -475,10 +477,83 @@ func TestInternalClientAdapterWithClientStreamInterceptors(t *testing.T) {
 	require.Equal(t, len(internal.rangeFeedEvents)+1, s.recvCount)
 }
 
+// Test that a server stream interceptor can wrap the ServerStream when the
+// internalClientAdapter is used.
+func TestInternalClientAdapterWithServerStreamInterceptors(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	ctx := context.Background()
+
+	stopper := stop.NewStopper()
+	defer stopper.Stop(context.Background())
+	stopper.SetTracer(tracing.NewTracer())
+
+	// Can't be zero because that'd be an empty offset.  !!!
+	clock := hlc.NewClock(timeutil.NewManualTime(timeutil.Unix(0, 1)), time.Nanosecond)
+
+	serverCtx := newTestContext(uuid.MakeV4(), clock, stopper)
+	serverCtx.Config.Addr = "127.0.0.1:9999"
+	serverCtx.Config.AdvertiseAddr = "127.0.0.1:8888"
+	serverCtx.NodeID.Set(context.Background(), 1)
+
+	_ /* server */, serverInterceptors := NewServerEx(serverCtx)
+
+	const int1Name = "interceptor 1"
+	serverInterceptors.StreamInterceptors = append(serverInterceptors.StreamInterceptors,
+		func(
+			srv interface{}, ss grpc.ServerStream, info *grpc.StreamServerInfo, handler grpc.StreamHandler,
+		) error {
+			serverStream := &testServerStream{name: "interceptor 1", inner: ss}
+			return handler(srv, serverStream)
+		})
+	var secondInterceptorWrapped grpc.ServerStream
+	const int2Name = "interceptor 2"
+	serverInterceptors.StreamInterceptors = append(serverInterceptors.StreamInterceptors,
+		func(
+			srv interface{}, ss grpc.ServerStream, info *grpc.StreamServerInfo, handler grpc.StreamHandler,
+		) error {
+			secondInterceptorWrapped = ss
+			serverStream := &testServerStream{name: int2Name, inner: ss}
+			return handler(srv, serverStream)
+		})
+
+	internal := &internalServer{rangeFeedEvents: []roachpb.RangeFeedEvent{{}, {}}}
+	serverCtx.SetLocalInternalServer(internal, serverInterceptors, ClientInterceptorInfo{})
+	ic := serverCtx.GetLocalInternalClientForAddr(serverCtx.Config.AdvertiseAddr, 1)
+	lic, ok := ic.(internalClientAdapter)
+	require.True(t, ok)
+	require.Equal(t, internal, lic.server)
+
+	stream, err := lic.RangeFeed(ctx, &roachpb.RangeFeedRequest{})
+	require.NoError(t, err)
+
+	// Consume the stream. This will synchronize with the server RPC handler
+	// goroutine, ensuring that the server-side interceptors run.
+	for {
+		_, err := stream.Recv()
+		if err == io.EOF {
+			break
+		}
+		require.NoError(t, err)
+	}
+
+	require.IsType(t, &testServerStream{}, secondInterceptorWrapped)
+	require.Equal(t, int1Name, secondInterceptorWrapped.(*testServerStream).name)
+	require.IsType(t, rangeFeedServerAdapter{}, internal.serverStream)
+	ss := internal.serverStream.(rangeFeedServerAdapter).ServerStream
+	require.IsType(t, &testServerStream{}, ss)
+	topStream := ss.(*testServerStream)
+	require.Equal(t, int2Name, topStream.name)
+	require.IsType(t, &testServerStream{}, topStream.inner)
+	bottomStream := topStream.inner.(*testServerStream)
+	require.Equal(t, int1Name, bottomStream.name)
+}
+
 type testClientStream struct {
 	inner     grpc.ClientStream
 	recvCount int
 }
+
+var _ grpc.ClientStream = &testClientStream{}
 
 func (t *testClientStream) Header() (metadata.MD, error) {
 	return t.inner.Header()
@@ -505,7 +580,36 @@ func (t *testClientStream) RecvMsg(m interface{}) error {
 	return t.inner.RecvMsg(m)
 }
 
-var _ grpc.ClientStream = &testClientStream{}
+type testServerStream struct {
+	name  string
+	inner grpc.ServerStream
+}
+
+var _ grpc.ServerStream = &testServerStream{}
+
+func (t testServerStream) SetHeader(md metadata.MD) error {
+	return t.inner.SetHeader(md)
+}
+
+func (t testServerStream) SendHeader(md metadata.MD) error {
+	return t.inner.SendHeader(md)
+}
+
+func (t testServerStream) SetTrailer(md metadata.MD) {
+	t.inner.SetTrailer(md)
+}
+
+func (t testServerStream) Context() context.Context {
+	return t.inner.Context()
+}
+
+func (t testServerStream) SendMsg(m interface{}) error {
+	return t.inner.SendMsg(m)
+}
+
+func (t testServerStream) RecvMsg(m interface{}) error {
+	return t.inner.RecvMsg(m)
+}
 
 func BenchmarkInternalClientAdapter(b *testing.B) {
 	ctx := context.Background()

